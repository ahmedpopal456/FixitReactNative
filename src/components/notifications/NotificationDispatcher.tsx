import React from 'react';
import {
  connect,
  rootContext,
  notificationActions
} from 'fixit-common-data-store';
import NotificationFactory from '../../factories/notificationFactory';
import {NotificationProps} from '../../models/notifications/NotificationProps';
import {FirebaseMessagingTypes} from '@react-native-firebase/messaging';

class NotificationDispatcher extends React.Component<any> {

  render() {
    return <NotificationFactory {...this._buildChildProps()} />;
  }

  private _buildChildProps(): NotificationProps {
    const message = this.props.messages[this.props.messages?.length - 1];
    const notificationType = +message?.data?.type;

    return {
      message: message,
      type: notificationType,
      onDismissNotification: this.props.onDismissNotification,
    };
  }
}

const mapStateToProps = (state: {notifications: {messages: FirebaseMessagingTypes.RemoteMessage[]}}) => {
  return {
    messages: state.notifications.messages,
  };
};

const mapDispatchToProps = (dispatch: Function) => {
  return {
    onDismissNotification: (id: string) => {
      dispatch(notificationActions.default.dismissNotification(id));
    },
  };
};

//@ts-ignore
export default connect(mapStateToProps, mapDispatchToProps, null, {
  context: rootContext,
})(NotificationDispatcher);
