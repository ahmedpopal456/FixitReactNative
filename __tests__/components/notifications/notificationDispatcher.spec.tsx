import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import { Provider, notificationActions, rootContext } from 'fixit-common-data-store';
import configureStore from 'redux-mock-store';
import { render, queryByAttribute } from 'react-testing-library';
import { Button } from 'fixit-common-ui';
import NotificationDispatcher from '../../../src/components/notifications/NotificationDispatcher';

jest.useFakeTimers();
const mockStore = configureStore();
describe('NotificationDispatcher', () => {
  let store: any;
  let component: any;
  const initialState = {
    notifications: {
      messages: [{
        messageId: 'test-notif-id',
        notification: {
          title: 'test-1',
        },
      }],
    },
  };

  beforeEach(() => {
    store = mockStore(initialState);
    store.dispatch = jest.fn();
    component = renderer.create(
      <Provider store={store} context={rootContext}>
        <NotificationDispatcher />
      </Provider>,
    );
  });

  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should dispatch dismissNotification on button press', () => {
    renderer.act(() => {
      renderer.act(() => {
        component.root.findByType(Button).props.onPress();
      });

      expect(store.dispatch).toHaveBeenCalledTimes(1);
      expect(store.dispatch).toHaveBeenCalledWith(
        notificationActions.default.dismissNotification('test-notif-id'),
      );
    });
  });
});
