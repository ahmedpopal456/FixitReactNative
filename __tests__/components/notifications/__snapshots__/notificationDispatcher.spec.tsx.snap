// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationDispatcher should render with given state from Redux store 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={true}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "marginTop": 22,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 20,
          "elevation": 5,
          "margin": 20,
          "padding": 35,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 4,
        }
      }
    >
      <Text
        style={
          Object {
            "marginBottom": 15,
            "textAlign": "center",
          }
        }
      >
        FixClientRequestNotification
      </Text>
      <Text
        style={
          Object {
            "fontSize": 20,
            "fontWeight": "800",
            "marginBottom": 15,
            "textAlign": "center",
          }
        }
      >
        test-1
      </Text>
      <Text
        style={
          Object {
            "marginBottom": 15,
            "textAlign": "center",
          }
        }
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "flex-start",
            "backgroundColor": "#FFD14A",
            "borderColor": "#FFD14A",
            "borderRadius": 3,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 5,
            "marginRight": 5,
            "marginTop": 5,
            "opacity": 1,
            "paddingBottom": 7,
            "paddingLeft": 7,
            "paddingRight": 7,
            "paddingTop": 7,
          }
        }
        testID="test-dismiss"
      >
        <View
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          <Text
            color="primary"
            fontSize={17}
            fontWeight={400}
            justification="center"
            padding={6}
            style={
              Array [
                Object {
                  "color": "#33343E",
                  "flexDirection": "row",
                  "fontFamily": "Raleway",
                  "fontSize": 17,
                  "fontWeight": "400",
                  "paddingBottom": 6,
                  "paddingLeft": 6,
                  "paddingRight": 6,
                  "paddingTop": 6,
                  "textAlign": "center",
                },
              ]
            }
          >
            Dismiss
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
